lte/gateway/python/magma/pipelined/ng_manager/session_state_manager.py

```diff
@@ -0,0 +1,266 @@
+            return CauseIE.REQUEST_ACCEPTED
+
+        # Check if the new session operation is without any pdr group
+        if len(new_session.set_gr_pdr) == 0:
+            return CauseIE.MANDATORY_IE_MISSING
+
+        # Check if the new session operation is without any far group
+        if len(new_session.set_gr_far) == 0:
+            return CauseIE.MANDATORY_IE_MISSING
+
+        return CauseIE.REQUEST_ACCEPTED
+
+    # Will return List of PDRs with Embed FARs
+    def translate_message_to_rules (self, new_session, process_pdr_rules, context_response):
+
+        # Parse and translate FAR rules
+        new_far_rules = OrderedDict()
+
+        #Create FAR rules
+        far_parse = self._far_create_rule_group (new_session.set_gr_far,
+                                                 new_far_rules)
+
+        if far_parse.cause != CauseIE.REQUEST_ACCEPTED:
+            context_response.cause_info.cause_ie = far_parse.cause
+            context_response.failure_rule_id.extend([
+                        FailureRuleInformation(pdr=OffendingIE(identifier=far_parse.identfier,
+                                               version=far_parse.sess_ver_no))])
+            return
+
+        #Create PDR rules
+        pdr_parse = self._pdr_create_rule_group (new_session, new_far_rules,
+                                                 process_pdr_rules)
+
+        if pdr_parse.cause != CauseIE.REQUEST_ACCEPTED:
+            context_response.cause_info.cause_ie = pdr_parse.cause
+            context_response.failure_rule_id.extend([
+                         FailureRuleInformation(pdr=OffendingIE(identifier=pdr_parse.identfier,
+                                                                version=pdr_parse.sess_ver_no))])
+
+        if new_session.state.state == Fsm_state.ACTIVE:
+            self.calculate_missing_pdrs (new_session.seid, process_pdr_rules) 
+
+    def process_session_message (self, new_session, process_pdr_rules):
+        """
+        Process the messages recevied from session. Return True
+        if parsing is successfull.
+        """
+
+        # Assume things are green
+        context_response =\
+             UPFSessionContextState(cause_info=CauseIE(cause_ie=CauseIE.REQUEST_ACCEPTED),
+                                    session_snaphot=UPFSessionState(seid=new_session.seid,
+                                                                    sess_ver_no=new_session.sess_ver_no))
+
+        context_response.cause_info.cause_ie = \
+                  self.validate_session_msg (new_session)
+        if context_response.cause_info.cause_ie != CauseIE.REQUEST_ACCEPTED:
+            return context_response
+
+        self.translate_message_to_rules (new_session, process_pdr_rules, context_response)
+
+        return context_response
+
+    def _monitor(self, poll_interval=5):
+        """
+        Main thread that sends a stats request at the configured interval in
```

Comment:
I don't think this actually sends anything

Reply:
This function is now changed a bit...It needs to be triggered from poll_stats of enforcement_stats.py. Will address once the proto files are baselined.